File: Packages/StackEngineCore/Sources/StackEngineCore/Luma.swift
import simd

public enum Luma {
    /// Rec.709 luminance of a single linear-RGB pixel.
    @inline(__always) public static func rec709(_ p: SIMD3<Float>) -> Float {
        0.2126 * p.x + 0.7152 * p.y + 0.0722 * p.z
    }

    /// Rec.709 luminance of each pixel.
    public static func luminance(_ img: PixelImage) -> [Float] {
        img.pixels.map { rec709($0) }
    }

    /// Sharpness = sum of |Laplacian| over the luminance image (higher = sharper).
    public static func sharpness(_ img: PixelImage) -> Float {
        sharpness(of: luminance(img), width: img.width, height: img.height)
    }

    /// Sharpness over a precomputed luminance buffer, so callers can reuse the buffer.
    static func sharpness(of l: [Float], width w: Int, height h: Int) -> Float {
        @inline(__always) func at(_ x: Int, _ y: Int) -> Float {
            l[min(max(y, 0), h - 1) * w + min(max(x, 0), w - 1)]
        }
        var s: Float = 0
        for y in 0..<h { for x in 0..<w {
            let lap = at(x-1, y) + at(x+1, y) + at(x, y-1) + at(x, y+1) - 4 * at(x, y)
            s += abs(lap)
        }}
        return s
    }
}

File: android/stackengine/src/main/kotlin/com/jdpneto/stackengine/Luma.kt
package com.jdpneto.stackengine

import kotlin.math.abs

/** Rec.709 luminance utilities. */
object Luma {

    /** Rec.709 luminance of a single linear-RGB pixel. */
    fun rec709(p: Vec3): Float = 0.2126f * p.x + 0.7152f * p.y + 0.0722f * p.z

    /** Rec.709 luminance of each pixel. */
    fun luminance(img: PixelImage): FloatArray {
        val n = img.pixelCount
        val out = FloatArray(n)
        for (i in 0 until n) {
            val base = i * 3
            out[i] = 0.2126f * img.pixels[base] + 0.7152f * img.pixels[base + 1] + 0.0722f * img.pixels[base + 2]
        }
        return out
    }

    /** Sharpness = sum of |Laplacian| over the luminance image (higher = sharper). */
    fun sharpness(img: PixelImage): Float =
        sharpness(luminance(img), img.width, img.height)

    /**
     * Sharpness over a precomputed luminance buffer, so callers can reuse the buffer.
     * Uses edge-clamped border handling.
     */
    fun sharpness(l: FloatArray, w: Int, h: Int): Float {
        fun at(x: Int, y: Int): Float {
            val cx = x.coerceIn(0, w - 1)
            val cy = y.coerceIn(0, h - 1)
            return l[cy * w + cx]
        }
        var s = 0f
        for (y in 0 until h) {
            for (x in 0 until w) {
                val lap = at(x - 1, y) + at(x + 1, y) + at(x, y - 1) + at(x, y + 1) - 4f * at(x, y)
                s += abs(lap)
            }
        }
        return s
    }
}
